### DISCUSSION TITLE
Help Needed: Camera Calibration for Tello Drone Adaptation

### DISCUSSION THREAD

**Discussion by d4saher** [jyjblrd/Mocap-Drones#50] [Q&A]
Hi jyjblrd,

First of all, thank you so much for sharing your code and making it available to the community. I really appreciate it as I will be using it in my final degree project as a ground truth.

I wanted to let you know that I have successfully got the system up and running. I am currently working on adapting the project for the Tello drone to simplify the drone part. Additionally, I would like to contribute to the project by sharing the code I write and some 3D pieces that I have designed. The pieces are basically a wall mount for the cameras and a support for mounting the IR LEDs on the Tello Drone EDU. You can find them on my [fork repository](https://github.com/d4saher/Low-Cost-Mocap/tree/tello) of your project, on the tello branch.

I have also written a step-by-step guide to get the project running on a Rock (Pi) 4 c+. I wrote this guide for myself, but maybe someone else will find it useful. It can be found [here](https://github.com/d4saher/Low-Cost-Mocap/blob/tello/step_by_step.md).

However, at this point, I am encountering an issue with camera pose calibration. I am able to get points (20 to 30 as you mentioned [here](https://github.com/jyjblrd/Low-Cost-Mocap/discussions/23#discussioncomment-8118097)), but when I run the camera pose calibration, it doesn't reach a solution. Using only two cameras seems to work better. Do I need to do this in pairs as you say in your video, or does the backend handle it? I followed the same method you used with the PS3 Eye cameras, removing the IR filter. Do I need to calculate the intrinsic matrix anyway, or is it enough to use the ones provided in your code?

Thank you once again for your amazing work!

Best regards,
D4saher

**Comment by jyjblrd** [comment]
Hi,

Thank you, I appreciate the contributions and guide.

The code automatically does the camera pose calibration in pairs. I would recommend generating new camera parameters for your cameras, as it can make a big difference. Instructions to do so are here: https://github.com/jyjblrd/Low-Cost-Mocap/discussions/11#discussioncomment-9380283 

The cameras should also be ordered in a clockwise/anti-clockwise pattern, so that adjacent cameras are also adjacent on the dashboard.

**Reply by d4saher** [reply]
Hi again,

Thanks for your reply.

It turns out the order of the cameras was indeed related to my problem. I hadn't considered that. Now that I have them in the correct order, I am attempting to calibrate them following the instructions, but I'm still facing some issues. I think it would be best to discuss these in [this discussion](https://github.com/jyjblrd/Low-Cost-Mocap/discussions/11). Specifically, I can't download the chessboard that gumby0q shared. I tried using another one, but I'm unsure if the square size in mm matters.

Thank you again for your help!

**Reply by d4saher** [reply]
Sorry to bother you again, but I'm facing an issue acquiring the floor. I could set the scale without problems using 2 IR leds separated by 15cm, and setting the origin was easy too. The problem is that now I capture around 4 points, each at the corners of the room, and I can see that they draw a plane. However, when I hit Acquire Floor on the chart, it looks like one of the cameras stands at the origin, as you can see in the picture.
![Captura desde 2024-06-12 21-18-48](https://github.com/jyjblrd/Low-Cost-Mocap/assets/113984123/12f6399d-7b5e-4ae9-9208-7a78978bfcb2)

This other image is before pressing Acquire Floor:
![Captura desde 2024-06-12 21-03-18](https://github.com/jyjblrd/Low-Cost-Mocap/assets/113984123/29c7d58a-407c-41e8-b2ca-fdd24d8ce325)

I am trying to understand how the code works, but my knowledge about vector spaces is limited. It looks like the normal vector is being calculated fine and so is the To World Matrix. I don't know if there is something else that I am missing. If you could orient me in the right direction, I would appreciate it.

Thank you for your assistance!

**Reply by jyjblrd** [reply]
Have you tried setting the floor before setting the origin?

You can also capture more that 4 points for the floor, you can "paint" the floor by dragging a marker across the floor.

**Reply by d4saher** [reply]
Yes, I also tried to do it before scale and repeated the entire process, but it always paints a camera on the origin and doesn't align the floor correctly. Tomorrow I will try to create manually some matrix to see how it applies them to cameras. Is there something else that I could try?

Thanks

**Reply by d4saher** [reply]
Hi,

This is what I got so far:
In the acquire_floor function, I am not sure, but I think there is a mistake with the up_normal. Originally, it is up_normal = np.array([[0],[0],[1]], dtype=np.float32). But since three.js uses a right-handed coordinate system, I think it should be up_normal = np.array([[0],[1],[0]], dtype=np.float32).

With that change, I got the cameras almost perpendicular to the floor, so I also had to apply a 90-degree rotation to the x-axis:
```
rotationMatrixX = np.array([
    [1, 0, 0],
    [0, 0, -1],
    [0, 1, 0]
])
```
Now it seems almost correct to me:
![Captura desde 2024-06-15 21-01-06](https://github.com/jyjblrd/Low-Cost-Mocap/assets/113984123/c849c55c-2773-49c1-8a5c-342ffc0351b4)

Need to check with cameras on Monday as I don't have them with me this weekend.

What do you think? Does it make sense?

**Reply by jyjblrd** [reply]
If I'm honest, the calibration stuff is not very well written and is quite hacky, especially since threejs uses a different coordinate system to the backend. I would like to clean it up one day, or perhaps someone else can. 

Have you had any success?

**Reply by nbpilot007** [reply]
> Hi,
> 
> This is what I got so far: In the acquire_floor function, I am not sure, but I think there is a mistake with the up_normal. Originally, it is up_normal = np.array([[0],[0],[1]], dtype=np.float32). But since three.js uses a right-handed coordinate system, I think it should be up_normal = np.array([[0],[1],[0]], dtype=np.float32).
> 
> With that change, I got the cameras almost perpendicular to the floor, so I also had to apply a 90-degree rotation to the x-axis:
> 
> ```
> rotationMatrixX = np.array([
>     [1, 0, 0],
>     [0, 0, -1],
>     [0, 1, 0]
> ])
> ```
> 
> Now it seems almost correct to me: ![Captura desde 2024-06-15 21-01-06](https://private-user-images.githubusercontent.com/113984123/340021148-c849c55c-2773-49c1-8a5c-342ffc0351b4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk2NDUwNDQsIm5iZiI6MTcxOTY0NDc0NCwicGF0aCI6Ii8xMTM5ODQxMjMvMzQwMDIxMTQ4LWM4NDljNTVjLTI3NzMtNDljMS04YTVjLTM0MmZmYzAzNTFiNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyOVQwNzA1NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZWMzZmE0ZmVlMDFmYjkzYTExMTZiMWMxNmNmNGM1YjdkYmM2YjhhZGJmYWUyOWVlNDM5NjRmODE1YzRlMzYzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.QQUdYlz47XjKDk4BASoGapqDNcAaihBWhvvmOIr45qQ)
> 
> Need to check with cameras on Monday as I don't have them with me this weekend.
> 
> What do you think? Does it make sense?

I'm facing the same issue. Did it solve your problem after changing the matrix? If yes, can you please share the entire modified function where you made the changes? Thanks

**Reply by d4saher** [reply]
> If I'm honest, the calibration stuff is not very well written and is quite hacky, especially since threejs uses a different coordinate system to the backend. I would like to clean it up one day, or perhaps someone else can.
> 
> Have you had any success?

Hi jyjblrd,

I apologize for the delay in responding; I've been away for a couple of weeks. Unfortunately, the last solution I proposed did not work as expected. I'm still investigating the cause of the issue. Additionally, I'm working on cleaning up the code and adapting it to better fit programming design patterns.

I have a question: when plotting the captured points on the graph, should they be drawn in the same direction as in reality? Right now, if we look at the graph as if it were the room from above and I draw a stroke with the marker from left to right, it gets plotted from right to left on the graph. I’m not sure if this might be affecting something.

I'll keep you updated on my progress.

Thanks,

**Reply by d4saher** [reply]
> > Hi,
> > This is what I got so far: In the acquire_floor function, I am not sure, but I think there is a mistake with the up_normal. Originally, it is up_normal = np.array([[0],[0],[1]], dtype=np.float32). But since three.js uses a right-handed coordinate system, I think it should be up_normal = np.array([[0],[1],[0]], dtype=np.float32).
> > With that change, I got the cameras almost perpendicular to the floor, so I also had to apply a 90-degree rotation to the x-axis:
> > ```
> > rotationMatrixX = np.array([
> >     [1, 0, 0],
> >     [0, 0, -1],
> >     [0, 1, 0]
> > ])
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Now it seems almost correct to me: ![Captura desde 2024-06-15 21-01-06](https://private-user-images.githubusercontent.com/113984123/340021148-c849c55c-2773-49c1-8a5c-342ffc0351b4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk2NDUwNDQsIm5iZiI6MTcxOTY0NDc0NCwicGF0aCI6Ii8xMTM5ODQxMjMvMzQwMDIxMTQ4LWM4NDljNTVjLTI3NzMtNDljMS04YTVjLTM0MmZmYzAzNTFiNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyOVQwNzA1NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZWMzZmE0ZmVlMDFmYjkzYTExMTZiMWMxNmNmNGM1YjdkYmM2YjhhZGJmYWUyOWVlNDM5NjRmODE1YzRlMzYzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.QQUdYlz47XjKDk4BASoGapqDNcAaihBWhvvmOIr45qQ)
> > Need to check with cameras on Monday as I don't have them with me this weekend.
> > What do you think? Does it make sense?
> 
> I'm facing the same issue. Did it solve your problem after changing the matrix? If yes, can you please share the entire modified function where you made the changes? Thanks

Hi, I am still having the same problem, trying to figure out what is the problem...

**Comment by d4saher** [comment]
Hi,

I am still trying to figure out where the problem is, but meanwhile, I have modified the program to allow manual adjustment of the scene with some buttons. I will be unavailable this weekend, so I won't be able to test if it works yet. If someone wants to try, the changes are on my fork repository on the dev branch: https://github.com/d4saher/Low-Cost-Mocap.git

Note that I am working on splitting the frontend into views to make the code more readable and modular, but I haven't added the drone control parts yet.

**Reply by jyjblrd** [reply]
Sorry, I have also been on holiday the last few weeks. I am hoping to refine the code for this project too at some point, as the calibration step is quite confusing currently (both in the code and in practice...)

I'm not sure what your specific problem is, it has worked for me and others so perhaps it is something to do with your setup? Perhaps you can send some pictures?

**Reply by d4saher** [reply]
Don't worry, I greatly appreciate your effort and dedication.

I'll try to send some photos this afternoon. Meanwhile, I had a thought: maybe in the backend function that calculates the floor, the calculated rotation isn't being applied to the old toWorldCoordsMatrix. Could this be the issue?

```
@socketio.on("acquire-floor")
def acquire_floor(data):
    cameras = Cameras.instance()
    object_points = data["objectPoints"]
    object_points = np.array([item for sublist in object_points for item in sublist])

    tmp_A = []
    tmp_b = []
    for i in range(len(object_points)):
        tmp_A.append([object_points[i,0], object_points[i,1], 1])
        tmp_b.append(object_points[i,2])
    b = np.matrix(tmp_b).T
    A = np.matrix(tmp_A)

    fit, residual, rnk, s = linalg.lstsq(A, b)
    fit = fit.T[0]

    plane_normal = np.array([[fit[0]], [fit[1]], [-1]])
    plane_normal = plane_normal / linalg.norm(plane_normal)
    up_normal = np.array([[0],[0],[1]], dtype=np.float32)

    plane = np.array([fit[0], fit[1], -1, fit[2]])

    # https://math.stackexchange.com/a/897677/1012327
    G = np.array([
        [np.dot(plane_normal.T,up_normal)[0][0], -linalg.norm(np.cross(plane_normal.T[0],up_normal.T[0])), 0],
        [linalg.norm(np.cross(plane_normal.T[0],up_normal.T[0])), np.dot(plane_normal.T,up_normal)[0][0], 0],
        [0, 0, 1]
    ])
    F = np.array([plane_normal.T[0], ((up_normal-np.dot(plane_normal.T,up_normal)[0][0]*plane_normal)/linalg.norm((up_normal-np.dot(plane_normal.T,up_normal)[0][0]*plane_normal))).T[0], np.cross(up_normal.T[0],plane_normal.T[0])]).T
    R = F @ G @ linalg.inv(F)

    R = R @ [[1,0,0],[0,-1,0],[0,0,1]] # i dont fucking know why

    cameras.to_world_coords_matrix = np.array(np.vstack((np.c_[R, [0,0,0]], [[0,0,0,1]])))

    socketio.emit("to-world-coords-matrix", {"to_world_coords_matrix": cameras.to_world_coords_matrix.tolist()})

```

cameras.to_world_coords_matrix = np.array(np.vstack((np.c_[R, [0,0,0]], [[0,0,0,1]]))) Here, the matrix is replaced, but the rotation is not applied to the old one. Shouldn't it be applied?

**Reply by d4saher** [reply]
Here are some photos of my setup. As you can see I have 4 cams each one on a corner of the room:

![IMG_20240715_182249.jpg](https://github.com/user-attachments/assets/37706735-939a-4fe0-af66-69c2e5f1ae43)

![IMG_20240715_182329.jpg](https://github.com/user-attachments/assets/6a8f5b24-b036-4fca-b863-0581dbb95d58)
